### PR TITLE
[hail] variation of `hl.float`/`hl.int` that returns null on invalid string

### LINKED ISSUE #7408: ability to cast to float with null if not a float
some version of `hl.float('test')` that returns `None` if it can't be coerced (rather than throw `NumberFormatException`) would be super useful. or a regex check could also accomplish this (with a `case`).

### PR DESCRIPTION
**Pull request by iitalics**
fixes #7408 

~~stacked on #7452~~

adds new toplevel hail functions:
  * `parse_float`, `parse_float32`, `parse_float64`
  * `parse_int`, `parse_int32`, `parse_int64`

all of the above accept only `tstr` expressions, and return missing for invalid numbers.

```python
>>> hl.eval(hl.parse_int("-5.3"))
>>> hl.eval(hl.parse_int("asdf"))
>>> hl.eval(hl.parse_int("-5"))
-5
>>> hl.eval(hl.parse_float("-NaM"))
>>> hl.eval(hl.parse_float("-NaN"))
nan
>>> hl.eval(hl.parse_float("5.3"))
5.3
```

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/functions/UtilFunctions.scala [hunk 1]
```diff
@@ -48,6 +48,25 @@ object UtilFunctions extends RegistryFunctions {
...
+
+  def isValidInt64(s: String): Boolean =
+    try { s.toLong; true } catch { case _: NumberFormatException => false }
+
+  def isValidFloat32(s: String): Boolean = parseSpecialNum(s) match {
+    case 0 => try { s.toFloat; true } catch { case _: NumberFormatException => false }
+    case _ => true
+  }
+
+  def isValidFloat64(s: String): Boolean = parseSpecialNum(s) match {
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/functions/UtilFunctions.scala]: You're parsing twice here -- once for the check and once for the parse. That's not ideal, but I do see the reason why it's not easy to do it just once in generated code without boxing.

**Comment by tpoterba** [comment]
(search for `delimit` and add in those places

File: hail/src/main/scala/is/hail/expr/ir/functions/UtilFunctions.scala (hunk 1 above)
Comment (iitalics) [reply hail/src/main/scala/is/hail/expr/ir/functions/UtilFunctions.scala]: yeah :I if we could catch exceptions in asm4s then that would be one nicer way to handle this.

**Comment by iitalics** [comment]
what the hell, bytecode verify errors from using locals? i saw another code-function creating locals so i assumed it was safe to do so, but apparently it isn't.

**Comment by danking** [comment]
I extremely like this. Eventually, I'd like to be able to write the table reading code in Python. That `import_matrix_table` only supports certain types is frustrating and it'd be easier to fix with more parsing functions like this!

**Pull request merged**